### PR TITLE
[PLC] [Builtins] Made type tags lists of ints

### PR DESCRIPTION
**Pull request by effectfully**
This should be more future-proof as the previous way we were encoding types from universes as tags did not allow to handle recursive types.

### CODE REVIEW

File: plutus-core/src/Language/PlutusCore/CBOR.hs [hunk 1]
```diff
@@ -94,9 +94,9 @@ decodeConstructorTag = decodeWord
 
 -- See Note [The G, the Tag and the Auto].
 instance Closed uni => Serialise (Some (TypeIn uni)) where
-    encode (Some (TypeIn uni)) = encodeConstructorTag . fromIntegral $ tagOf uni
+    encode (Some (TypeIn uni)) = encode . map (fromIntegral :: Int -> Word) $ encodeUni uni
```
Comment (michaelpj) [inline plutus-core/src/Language/PlutusCore/CBOR.hs:97]: This will make encoding constants a bit more expensive, but I think that's fine.

File: plutus-core/src/Language/PlutusCore/Parser.y [hunk 2]
```diff
@@ -174,35 +174,35 @@ getStaticBuiltinName = \case
...
-    "string"     -> Just $ tagOf DefaultUniString
-    "unit"       -> Just $ tagOf DefaultUniUnit
+encodeTyName :: T.Text -> Maybe [Int]
+encodeTyName = \case
+    "bool"       -> Just $ encodeUni DefaultUniBool
+    "bytestring" -> Just $ encodeUni DefaultUniByteString
+    "char"       -> Just $ encodeUni DefaultUniChar
+    "integer"    -> Just $ encodeUni DefaultUniInteger
+    "string"     -> Just $ encodeUni DefaultUniString
+    "unit"       -> Just $ encodeUni DefaultUniUnit
```
Comment (kwxm) [inline plutus-core/src/Language/PlutusCore/Parser.y:184]: Ultimately we don't want this to be coded into the Parser, but somehow done by something passed to the parser as a parameter, don't we?  Will that be in a separate PR?

File: plutus-core/src/Language/PlutusCore/CBOR.hs (hunk 1 above)
Comment (kwxm) [reply plutus-core/src/Language/PlutusCore/CBOR.hs:96]: If we really care about  saving the overhead incurred by encoding a list, can't we encode a single `Int` if the list has length 1 (which will always be the case at present) and encode the whole list otherwise, and then use `deserialiseOrFail` to deal with the two possibilities during decoding? Of course that'll make decoding a bit more expensive instead.  (Also, CBOR's probably going away.)

Comment (kwxm) [reply plutus-core/src/Language/PlutusCore/CBOR.hs:96]: (I'm not sure why that was `encodeConstructorTag` there: that's suppose to be used to encode Haskell constructors.)

File: plutus-core/src/Language/PlutusCore/Universe/Core.hs [hunk 3]
```diff
@@ -104,17 +101,28 @@ knownUniOf _ = knownUni
...
+-- @UList@ and @UInt@ constructors, respectively.
 class Closed uni where
     -- | A constrant for \"@constr a@ holds for any @a@ from @uni@\".
     type Everywhere uni (constr :: GHC.Type -> Constraint) :: Constraint
 
-    -- | Get the 'Int' tag of a type from @uni@. The opposite of 'uniAt'.
-    tagOf :: uni a -> Int
+    -- | Encode a type as a sequence of 'Int' tags.
+    -- The opposite of 'decodeUni'.
+    encodeUni :: uni a -> [Int]
```
Comment (kwxm) [inline plutus-core/src/Language/PlutusCore/Universe/Core.hs:121]: Perhaps `NonEmpty Int`?

**Review by kwxm — APPROVED** [review]
Looks fairly harmless (famous last words ...).

**Review by kwxm — APPROVED** [review]

File: plutus-core/src/Language/PlutusCore/Universe/Core.hs (hunk 3 above)
Comment (effectfully) [reply plutus-core/src/Language/PlutusCore/Universe/Core.hs:121]: Then `NonEmpty Natural`. I don't feel like we need any safety here and `[]` has way better ergonomics than `NonEmpty`.

File: plutus-core/src/Language/PlutusCore/Parser.y (hunk 2 above)
Comment (effectfully) [reply plutus-core/src/Language/PlutusCore/Parser.y:184]: Yes, that's for a different PR.

**Comment by michaelpj** [comment]
You'll need to accept the new output for the `plutus-use-cases` tests, since this has changed Tx hashes. `cabal test plutus-use-cases --test-options=--accept` will do it.

**Comment by effectfully** [comment]
> You'll need to accept the new output for the `plutus-use-cases` tests, since this has changed Tx hashes.

Thanks, I just hadn't look in what CI reports.

File: plutus-core/src/Language/PlutusCore/CBOR.hs (hunk 1 above)
Comment (effectfully) [reply plutus-core/src/Language/PlutusCore/CBOR.hs:96]: > If we really care about saving the overhead incurred by encoding a list, can't we encode a single Int if the list has length 1

Would that be more optimal? It seems like the overhead of distinguish between this two cases is not very much different from just serializing the length of the list (which is what CBOR does, I believe).

Comment (michaelpj) [reply plutus-core/src/Language/PlutusCore/CBOR.hs:96]: Yeah, I think this is probably not too bad. And there are comparatively few constants in a program compared to other AST nodes.

**Comment by michaelpj** [comment]
... and you need to do that after rebasing on master so you don't conflict with the changes there.

**Comment by michaelpj** [comment]
:+1:

File: plutus-core/src/Language/PlutusCore/CBOR.hs (hunk 1 above)
Comment (kwxm) [reply plutus-core/src/Language/PlutusCore/CBOR.hs:96]: > It seems like the overhead of distinguish between this two cases is not very much different from just serializing the length of the list (which is what CBOR does, I believe).

Just for the sake of nitpicking, that's not actually what CBOR does (or more precisely, what CBORG does).  For a stream of items like a list or array, CBOR allows you to do two different things. You can emit a tag saying "here's a list of n items" and follow it with the encodings of the individual items; if the length is small you can fit it and the beginning-of-list marker into a single byte.  The encodings for small integers all fit into one byte as well, so you could encode `[2]` using two bytes.  However, you can also emit a tag saying "here's the start of a list of indefinite length", emit the encoding of the items, and then emit a tag saying "end of list", so `[2]` would require 3 bytes.  For nonempty lists, CBORG uses the latter strategy (or at least I'm pretty sure it does).  

So my suggestion would require 1 byte in to encode the common case of a one-element list with a single small entry, whereas encoding the whole list would require 3 bytes.  So if you had ten constants in a program you could save up to a massive 20 bytes!

As Michael says, we're probably not going to have many constants, so it's probably not worth the trouble.  If we actually cared (I don't) we could always do a before and after comparison to see if it really makes any difference, since I think my suggestion with the scheme in this PR will produce exactly the same encoding as the previous scheme for all of our existing examples.

[Back to my so-called holiday now...]

Comment (effectfully) [reply plutus-core/src/Language/PlutusCore/CBOR.hs:96]: @kwxm thank you, I just quickly looked on wiki what CBOR does, but did not look into the exact details of `cborg`.

Yeah, 3 bytes is definitely worse than 1 byte. Maybe we should do some actual measurements and adopt your strategy.

Comment (michaelpj) [reply plutus-core/src/Language/PlutusCore/CBOR.hs:96]: Note that we're almost certainly going to change our binary serialization format soon, so don't waste too much thought on CBOR.
